src/main/scala/com/srcd/swivel/SparkPrepDriver.scala

```diff
@@ -0,0 +1,415 @@
+    val vocab = SparkPrep.vocabFromDict(dict)
+    println(s"Done. ${dict.length} words found")
+
+    (vocab, dict)
+  }
+
+  /**
+    * Reads existing vocabulary
+    * @param vocabFile path to vocabulary file
+    */
+  def readVocab(vocabFile: String): (Map[String, Int], Seq[(String, Int)]) = {
+    println(s"Reading vocabulary from ${vocabFile}")
+    val dict = Source.fromFile(vocabFile)
+      .getLines
+      .toList
+      .zipWithIndex
+    val vocab = dict.toMap
+    println(s"Done. ${dict.length} words found")
+
+    (vocab, dict)
+  }
+
+  /**
+    * Converts array of tokens to Ids
+    *
+    * @param rdd
+    * @param wordToIdVar
+    * @return
+    */
+  def wordsToIds(rdd: RDD[Array[String]], wordToIdVar: Broadcast[Map[String, Int]]) = {
+    rdd.map(line => {
+        line.flatMap { // flat is important, skips OOV
+          wordToIdVar.value.get(_)
+        }
+      })
+  }
+
+
+  /**
+    * Transforms to Id each token, and list elements of co-ocurence matrix.
+    * Each matrix element (i,j) can be listed multiple times.
+    *
+    * @param rdd Tokenized lines of input
+    * @param windowSize size of co-occurrence window
+    * @param numShards numer of shards
+    * @param vocab word -> id
+    * @return co-occurrence matrix as ((i, j), weight)
+    */
+  def buildCooccurrenceMatrix(
+    rdd: RDD[Array[String]],
+    windowSize: Int,
+    numShards: Int,
+    vocab: Broadcast[Map[String, Int]]
+  ): RDD[((Int, Int), Double)] = {
+
+    val ids = wordsToIds(rdd, vocab)
+    val coocs = ids.flatMap(generateCooccurrence(_, windowSize))
+    coocs
+  }
+
+  def generateCooccurrence(ids: Array[Int], windowSize: Int) = {
+    val coocs = mutable.HashMap[(Int, Int), Double]().withDefaultValue(0)
+      0 until ids.length foreach { pos =>
+        val lid = ids(pos)
+        val windowExtent = Math.min(windowSize + 1, ids.length - pos)
+        1 until windowExtent foreach { off =>
+          val rid = ids(pos + off)
+          val count = 1.0 / off
+          val pair = (Math.min(lid, rid), Math.max(lid, rid))
+          coocs(pair) += count
+          coocs(pair.swap) += count
+        }
+        val pair = (lid, lid)
+        coocs(pair) += 1.0  // Add 1 (not 1/2 as before) since we do not output (a, b) and (b, a) separately
+      }
+      coocs
+  }
+
+  /**
+    * For given co-occurrence matrix, shards it to <bold>numShards * numShards</bold> pices
+    *
+    * @param coocs
+    * @param numShards number of shards along the dimension
+    * @return
+    */
+  def mergeCoocsAndShard(coocs: RDD[((Int, Int), Double)], numShards: Int): RDD[((Int, Int), Double)] = {
+    //TODO(bzz): if time permits, compare performance of 3 possible impls below
+    val shardedCoocs = coocs
+      //I. repartitionAndSort + smart manual reduce over .mapPartition(), takeing advantage of key order
+      //  .repartitionAndSortWithinPartitions(new ShardPartitioner(numShards))
```

Comment:
These again are breadcrumbs - there are 3 possible ways of implementing this function, all have performance consequences and all took time to discover.

Part of the future work for a person who support this project  will be going though TODOs, determining performance implications and picking the best option